### ISSUE TITLE
`metal::malloc` resource limit crash on long `lax.scan` / `lax.while_loop`

### ISSUE THREAD

**Issue by tsumme1** [tillahoffmann/jax-mps#134]
### Code to reproduce

```python
import jax
import jax.numpy as jnp
from jax import random
def make_lstm_params(key, input_dim, hidden_dim):
    keys = random.split(key, 8)
    # Wi, Wf, Wc, Wo (input projections) + Ui, Uf, Uc, Uo (recurrent projections)
    return tuple(random.normal(k, (input_dim if i < 4 else hidden_dim, hidden_dim)) * 0.01
                 for i, k in enumerate(keys))
def lstm_step(carry, x):
    (h, c), params = carry
    Wi, Wf, Wc, Wo, Ui, Uf, Uc, Uo = params
    i = jax.nn.sigmoid(x @ Wi + h @ Ui)
    f = jax.nn.sigmoid(x @ Wf + h @ Uf)
    g = jnp.tanh(x @ Wc + h @ Uc)
    o = jax.nn.sigmoid(x @ Wo + h @ Uo)
    c_new = f * c + i * g
    h_new = o * jnp.tanh(c_new)
    return ((h_new, c_new), params), h_new
hidden = 64
params = make_lstm_params(random.key(0), 32, hidden)
xs = jnp.ones((10_000, 32))
def loss_fn(params):
    init = ((jnp.zeros(hidden), jnp.zeros(hidden)), params)
    _, ys = jax.lax.scan(lstm_step, init, xs)
    return ys.sum()
jax.grad(loss_fn)(params)  # crashes: metal::malloc resource limit exceeded
```

## Problem
Long-running `lax.scan` / `lax.while_loop` (e.g., LSTM fwd+bwd with ≥10k timesteps) crash with:
```
[MPS ERROR] Exception dispatching stablehlo.dynamic_update_slice: [metal::malloc] Resource limit (499000) exceeded.
[MPS ERROR] stablehlo.while: failed to execute body region
```
## Current state
`HandleWhile` throws `CompileIncompatibleError` when called inside `mx::compile`, forcing the entire outer compilation to fall back to the interpreted MLIR walker. In this path, each loop iteration accumulates MLX graph nodes without evaluation. For long loops, unbounded graph growth exceeds Metal's ~500k allocation limit.

## Approach (I am already preparing a PR)
Replace the `CompileIncompatibleError` throw with a custom `WhileLoopPrimitive` — an `mlx::core::Primitive` subclass that is opaque to `mx::compile` but executes the loop internally:
- loop body wrapped via `mx::compile()` for GPU execution
- The primitive runs on the CPU stream for orchestration only (this does not cause any issues as far as I can tell) — the compiled body dispatches GPU kernels via the default GPU stream. Running on the GPU stream would deadlock (`eval()` inside a GPU eval callback re-enters the GPU scheduler).

### Version and runtime information

```text
jax:     0.9.2
jaxlib:  0.9.2
jax-mps: 0.9.14
Python:  3.13.12
macOS:   26.3.1 (25D2128)
Chip:    Apple M3 Max
Memory:  48 GB
Backend: mps
```

**Comment by tsumme1** [comment]
@tillahoffmann Good news and bad news.

Good news: Your buffer cache fix (PR #139) appears to have resolved the large memory allocation I was seeing during random_lax_test.py. I now have a version of the WhileLoopPrimitive that successfully runs `scripts/run_jax_tests.py`, and I will begin preparing a new PR.

Bad news: I noticed a significant performance regression compared to v0.9.13 in a GRU training benchmark I've been using. The regression affects the fully unrolled case (no while-loop generated), so it is likely related to one of the other recent changes. I've attached the benchmark below. If you run it with v0.9.13 and the current main, you should see a clear gap at h=1024 and h=2048.

[bench_rnn_training.py](https://github.com/user-attachments/files/26863776/bench_rnn_training.py)

Do you know which recent change might have introduced this? I would also be happy to open a separate issue and PR to add a few benchmarks that compare jax-mps performance directly against equivalent native MLX implementations. This would give us a ceiling to measure against and make it easier to spot regressions.

**Comment by imkow** [comment]
@tillahoffmann  @tsumme1 I noticed the same significant performance regression: 50%-70% slower on M5 Max 40GPU cores.

**Comment by tillahoffmann** [comment]
Thank you for sharing the information regarding the performance regression. I've oddly not been able to reproduce it on my machine (M4 MacBook Air). Are you able to reproduce it with this setup? https://gist.github.com/tillahoffmann/941323bf835f386b31bb0bee29bcf7aa

@tsumme1, does the performance regression disappear for you in #142?

**Comment by tsumme1** [comment]
I ran `bisect_perf.sh` on my M3 Max and confirmed the regression is **PR #139** (buffer cache cap). The 1 GiB hard-coded limit causes excessive buffer reallocation on higher core-count GPUs. That is why @imkow (M5 Max) and I see 30–70% regressions while your M4 Air stays within noise. The regression is present in #142, and workaround below works for it as well.

| commit | title | h1024_full (ms) | h2048_full (ms) | h1024_u1 (ms) | md5 |
|--------|-------|---:|---:|---:|:---:|
| a4f07f8 | Bump version to 0.9.14 | 71.82 | 169.19 | 1315.09 | 14fc885b |
| 1248f8d | StableHLO fusion framework (#136) | 72.90 | 171.13 | 1311.12 | 1325c32a |
| 694971e | Revert WhileLoopPrimitive (#137) | 72.85 | 172.53 | 1731.62 | d38c5525 |
| 6fcfca6 | MLIR threadpool + cache leaks (#138) | 73.82 | 171.66 | 1746.41 | bf7ae374 |
| **09692b0** | **Cap MLX buffer cache (#139)** | **93.58** | **288.30** | **2019.18** | **ced022c4** |
| a72010c | Hang-diagnosis tooling (#141) | 94.08 | 292.38 | 2098.64 | f3a45189 |

Workaround: `JAX_MPS_CACHE_LIMIT_BYTES=51539607552` (48 GiB) restores baseline performance on my device.

I suggest removing the default cap (or using MLX's default) and documenting JAX_MPS_CACHE_LIMIT_BYTES as an optional flag for scenarios like running the full JAX test suite where unbounded cache growth is a concern.

**Comment by tillahoffmann** [comment]
Thank you both for investigating. 1GB default removed in #143 with 1GB limit set for running the full test suite.
